lib/React.re

```diff
@@ -113,6 +111,41 @@ external useReducer:
   (('state, 'action) => 'state, 'state) => ('state, 'action => unit) =
   "useReducer";
 
+external useEffect: 
+  (unit => option(unit => unit)) => unit = 
+  "useEffect";
+
+// external useEffect0:
+//   (unit => option(unit => unit), Js.array_empty) => unit =
```

Comment:
I am not sure how to express an empty array for `useEffect0`. I did find [Js.array_empty](https://ocsigen.org/js_of_ocaml/3.1.0/api/Js) on the jsoo docs but how do I use it here? I tried `Js_of_ocaml.Js.array_empty` but to no avail.

Reply:
With BuckleScript bindings, which are kind of "inlined", we have to find a way to define constant values in the signature.

With the `external` approach of rroo that splits type definitions with actual implementation, we don't need to 🙂 We can just skip the dependencies param from the signature, and then in the `.js` implementation pass the empty array. ✨

Reply:
For an example of something similar, check how the `createFragment` function doesn't take any constant for the element type, just the array of children:
https://github.com/jchavarri/rroo/blob/7f5e9626742c9b2d8c602fb92c26544c7174d36e/lib/ReactDOM.re#L46

but later on, in the JS implementation, we read `React.Fragment` to pass it as the first param of `React.createElement`:

https://github.com/jchavarri/rroo/blob/7f5e9626742c9b2d8c602fb92c26544c7174d36e/lib/ReactDOM.js#L39-L43

Reply:
Ah I see. Would that mean the `useEffect` and `useEffect` would have the same external declarations? In that case would we not want `useEffect0`? Or do we want to keep it to keep API as close as possible to ReasoReact?

Reply:
Actually I am doing this for `useEffect` and it seems to be working:

`external useEffect0: (unit => option(unit => unit), array(unit)) => unit = "useEffect";`

Reply:
Thanks a lot for your work on this! 🙌 

> Ah I see. Would that mean the `useEffect` and `useEffect` would have the same external declarations? In that case would we not want `useEffect0`?

They would have the same type signatures but compiled down to different behavior, via the JavaScript binding code. The idea is:

- external `useEffect` -> function `useEffect` in `React.js` file -> calls `React.useEffect`
- external `useEffect0` -> function `useEffect0` in `React.js` file -> calls `React.useEffect` with an empty array

One could think of these two functions as `useEffectAlways` (for `useEffect`) and `useEffectOnce` (for `useEffect0`) if that makes sense. Although I would tend to think it's beneficial to keep the naming close to ReasonReact, for familiarity.

> Or do we want to keep it to keep API as close as possible to ReasonReact?

I think it's interesting to stay as close as ReasonReact as possible, but there is room to explore and tweak the APIs to make them more ergonomic, or the types more OCaml-idiomatic.

> Actually I am doing this for `useEffect` and it seems to be working

In this case, I feel more inclined to define `useEffect0` without the dependency param, because being a constant value, the `array(unit)` type might lead to confusion (what is an array of nothing? nothing 😄 ).

One important piece of context: In ReasonReact case, they chose those types because they had a very hard constraint on zero cost in the resulting bundle size, but in rroo that constraint does not exist 🙂